src/bp/core/services/middleware/state-manager.ts

```diff
@@ -1,35 +1,86 @@
 import * as sdk from 'botpress/sdk'
+import { KnexExtension } from 'common/knex'
 import { BotpressConfig } from 'core/config/botpress.config'
 import { ConfigProvider } from 'core/config/config-loader'
+import Database from 'core/database'
 import { createExpiry } from 'core/misc/expiry'
-import { inject, injectable } from 'inversify'
+import { inject, injectable, tagged } from 'inversify'
+import { Redis } from 'ioredis'
+import Knex from 'knex'
 import _ from 'lodash'
 import { Memoize } from 'lodash-decorators'
+import ms from 'ms'
 
 import { SessionRepository, UserRepository } from '../../repositories'
 import { TYPES } from '../../types'
 import { SessionIdFactory } from '../dialog/session/id-factory'
+import { JobService } from '../job-service'
 import { KeyValueStore } from '../kvs'
 
+const getRedisSessionKey = sessionId => `userstate_${sessionId}`
+const BATCH_SIZE = 100
+const MEMORY_PERSIST_INTERVAL = ms('3s')
+const REDIS_MEMORY_DURATION = ms('30s')
+
 @injectable()
 export class StateManager {
+  private _redisClient!: Redis
+  private batch!: { event: sdk.IO.IncomingEvent; ignoreContext?: boolean }[]
+  private knex!: Knex & KnexExtension
+  private currentPromise
+
   constructor(
+    @inject(TYPES.Logger)
+    @tagged('name', 'StateManager')
+    private logger: sdk.Logger,
     @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider,
     @inject(TYPES.UserRepository) private userRepo: UserRepository,
     @inject(TYPES.SessionRepository) private sessionRepo: SessionRepository,
-    @inject(TYPES.KeyValueStore) private kvs: KeyValueStore
+    @inject(TYPES.KeyValueStore) private kvs: KeyValueStore,
+    @inject(TYPES.Database) private database: Database,
+    @inject(TYPES.JobService) private jobService: JobService
   ) {}
 
+  public initialize() {
+    if (!process.CLUSTER_ENABLED) {
+      return
+    }
+
+    this.knex = this.database.knex
+    this.batch = []
+
+    const client = this.jobService.getRedisClient()
+    if (client) {
+      this._redisClient = client
+    }
+
+    setInterval(this._runTask, MEMORY_PERSIST_INTERVAL)
+  }
+
   private LAST_MESSAGES_HISTORY_COUNT = 5
   private BOT_GLOBAL_KEY = 'global'
 
   public async restore(event: sdk.IO.IncomingEvent) {
+    const sessionId = SessionIdFactory.createIdFromEvent(event)
+
+    if (process.CLUSTER_ENABLED) {
+      try {
+        const userState = await this._redisClient.get(getRedisSessionKey(sessionId))
+        if (userState) {
+          event.state = JSON.parse(userState)
+          event.state.__stacktrace = []
+          return
```

Comment:
I know we have this "early return" pretty much everywhere in the codebase but it becomes quite hard to read when you check the code quickly, you have to read the whole func to understand what's happening. We should try to avoid those kind of return statements and ensure we only use them as assertions (in the beginning) or actual return values (at the end).

Reply:
In this case, if it fails to restore the state from redis, I want it to read it from the database so there's no user impact.